### PR TITLE
Refactor documentation generation code. Extend fallback rules to include group.

### PR DESCRIPTION
**Pull request by cmnbroad**
Changes based on review of gatk usage.

### CODE REVIEW

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java [hunk 1]
```diff
@@ -307,35 +306,41 @@ private DocumentedFeatureObject getFeatureForClass(final Class<?> clazz) {
...
+                    clpGroup = clProps.programGroup().newInstance();
+                } catch (IllegalAccessException | InstantiationException e) {
                     logger.warn(
                             String.format("Cant instantiate program group class to retrieve summary for group %s for class %s",
                                     clProps.programGroup().getName(),
                                     clazz.getName()));
-                    groupSummary = "No group summary available";
+                }
+            }
+            if (summary.length() == 0) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: I always prefer `isEmpty` to `length == 0` checks.

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java [hunk 2]
```diff
@@ -307,35 +306,41 @@ private DocumentedFeatureObject getFeatureForClass(final Class<?> clazz) {
...
+                if (clpGroup != null) {
+                    summary = clProps.oneLineSummary();
+                }
+                if (summary.length() == 0) {
+                    logger.warn("No summary declared for: " + clazz.getCanonicalName() + ". Trying javadoc fallback.");
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: Do we no longer need the backup text of "No X available" or to worry about them being null?

File: src/main/java/org/broadinstitute/barclay/help/DefaultDocumentedFeatureHandler.java [hunk 3]
```diff
@@ -137,7 +133,7 @@ protected void addHighLevelBindings(
             final List<? extends CommandLinePluginDescriptor<?>> pluginDescriptors,
             final Map<String, Object> root)
     {
-        root.put("name", currentWorkUnit.classDoc.name());
+        root.put("name", currentWorkUnit.clazz.getSimpleName());
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DefaultDocumentedFeatureHandler.java]: Yes.

File: src/main/java/org/broadinstitute/barclay/help/DefaultDocumentedFeatureHandler.java [hunk 4]
```diff
@@ -256,11 +252,15 @@ protected void addCommandLineArgumentBindings(
     }
 
     private void addSummary(final Map<String, Object> root) {
-        final StringBuilder summaryBuilder = new StringBuilder();
-        for (final Tag tag : currentWorkUnit.classDoc.firstSentenceTags()) {
-            summaryBuilder.append(tag.text());
+        // If no summary was found from annotations, use the javadoc if there is any
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DefaultDocumentedFeatureHandler.java]: Agreed, though I think that the refactoring this code so badly needs should be done in a separate PR since it needs bigger modeling changes (i.e., DocumentedFeatureObject should probably go away completely, and DocWorkUnit should just contain the truth).

File: src/main/java/org/broadinstitute/barclay/help/DocumentedFeatureObject.java [hunk 5]
```diff
@@ -54,11 +54,14 @@ public DocumentedFeatureObject(
         this.groupSummary = groupSummary;
         this.extraDocs = extraDocs;
     }
-
     public Class<?> getClassToDoc() { return classToDoc; }
     public boolean enable() { return enable; }
-    public String summary() { return summary; }
+
+    public String getSummary() { return summary; }
+    public void putSummary(final String newSummary) { summary = newSummary; }
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocumentedFeatureObject.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java (hunk 1 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java (hunk 2 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: They're never null because the annotation attributes now all have default values of "". We no longer want to populate the strings with "No X declared" here since further on, when we have access to the javadoc, we'll use the empty string to determine if we should attempt to fall back to get a summary from that. I'll add a comment explaining that here. Ultimately this all needs refactoring.

Comment (lbergelson) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: Is it fair to say that we don't care about people who perversely over write the default with null?

**Comment by cmnbroad** [comment]
Added two more commits. The first is for the code review comments. The second is refactoring:

Removed classes:
- DocumentedFeatureObject (just use DocumentedFeature annotation directly)
- DocUtils (merged with DocletUtils)

Renamed classes:
- DocumentedFeatureHandler -> DocWorkUnitHandler
- DefaultDocumentedFeatureHandler -> DefaultDocWorkUnitHandler

Allow consumers to override creation of HelpDoclet, DocWorkUnit and DocWorkUnitHandler
Use builtin javadoc destination dir argument instead of custom argument

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java [hunk 6]
```diff
@@ -143,17 +147,16 @@ else if (!settingsDir.isDirectory())
      */
     public static int optionLength(final String option) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: I don't see this method used anywhere, here or in the gatk pull request.  Is it something we still need?  Is it part of the doclet contract?

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java [hunk 7]
```diff
@@ -225,125 +291,44 @@ private void processDocs(final RootDoc rootDoc) {
...
-        } else {
+            return clazz.getAnnotation(DocumentedFeature.class);
+        }
+        else {
             return null;
         }
     }
 
+    private CommandLineProgramProperties getCommandLineProgramPropertiesForClass(final Class<?> clazz) {
+        return clazz.isAnnotationPresent(CommandLineProgramProperties.class) ?
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: this is redundant i think, `getAnnotation()` returns null if it isn't present

File: src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java [hunk 8]
```diff
@@ -27,84 +27,189 @@
...
+    /**
+     * Return the description to be used for the work unit. We need to manually strip
+     * out any inline custom javdoc tags since we don't those in the summary.
+     *
+     * @param currentWorkUnit
+     * @return Descrition to be used or the work unit.
+     */
+    protected String getDescription(final DocWorkUnit currentWorkUnit) {
+        return Arrays.stream(currentWorkUnit.getClassDoc().inlineTags())
+                .filter(t -> getTagPrefix() == null ? true : !t.name().startsWith(getTagPrefix()))
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java]: Makes more sense. Done.

File: src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java [hunk 9]
```diff
@@ -27,84 +27,189 @@
...
+                logger.warn("No group summary declared for: " + workUnit.getClazz().getCanonicalName());
+                groupSummary = "";
+            }
+        }
+        return groupSummary;
+    }
+
+    /**
+     * Return the description to be used for the work unit. We need to manually strip
+     * out any inline custom javdoc tags since we don't those in the summary.
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java [hunk 10]
```diff
@@ -27,84 +27,189 @@
...
+     *
+     * @param currentWorkUnit
+     * @return Descrition to be used or the work unit.
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java [hunk 11]
```diff
@@ -285,38 +347,39 @@ private String getTagPrefix() {
     }
 
     // Add top level bindings for the default instances for each plugin descriptor
-    // NOTE: The default freemarker template provided by Barlcay has no references to plugins, since they're
+    // NOTE: The default Freemarker template provided by Barlcay has no references to plugins, since they're
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DefaultDocWorkUnitHandler.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java [hunk 12]
```diff
@@ -26,106 +26,186 @@
...
+     * Get the DocumentedFeature annotation object for this class.
+     * @return DocumentedFeature object. Will not be null.
      */
-    protected final ClassDoc classDoc;
+    public DocumentedFeature getDocumentedFeature() { return documentedFeature; }
+
     /**
-     * The documentedFeatureObject that lead to this Class being in Doc
+     * Get the JavDoc ClassDoc for this work unit.
+     * @return ClassDoc for this work unit. Will noe be null.
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java [hunk 13]
```diff
@@ -26,106 +26,186 @@
...
-    public String getTargetFileName() { return handler.getDestinationFilename(classDoc, clazz); }
+    /**
+     * The summary of the documentation object
+     */
+    public String getSummary() { return summary; }
+
+    public Class<?> getClazz() { return clazz; }
+
+    /**
+     * Populate the property map for this wokr unit by delegating to the documented feature handler for this work unit.
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java [hunk 14]
```diff
@@ -26,106 +26,186 @@
...
+     * @param featureMaps map of all features included in this javadoc run
+     * @param groupMaps map of all groups included in the javadoc run
+     */
+    public void processDoc(final List<Map<String, String>> featureMaps, final List<Map<String, String>> groupMaps) {
+        workUnitHandler.processWorkUnit(this, featureMaps, groupMaps);
+
+    };
+
+    /**
+     * Get the template to be used for this work unit. Delegates to th feature handler.
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java [hunk 15]
```diff
@@ -26,106 +26,186 @@
...
+     * Get the CommandLineProgramGroup object from the CommandLineProgramProperties of this work unit.
+     * @return CommandLineProgramGroup if the feature has one, otherwise null.
+     */
+    public CommandLineProgramGroup getCommandLineProgramGroup() {
+        if (commandLineProperties != null) {
+            try {
+                return commandLineProperties.programGroup().newInstance();
+            } catch (IllegalAccessException | InstantiationException e) {
+                logger.warn(
+                        String.format("Cant instantiate program group class to retrieve summary for group %s for class %s",
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocWorkUnit.java]: Done.

File: src/main/java/org/broadinstitute/barclay/help/DocWorkUnitHandler.java [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+import java.io.*;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Abstract base class for work unit handlers for docs. The DocWorkUnitHandler defines the template
+ * used for each documented feature, and populates the template property map for that template.
+ */
+public abstract class DocWorkUnitHandler {
+    private HelpDoclet doclet;
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocWorkUnitHandler.java]: Yes. Done.

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java (hunk 6 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: Yes, its called by javadoc.

File: src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java (hunk 7 above)
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/HelpDoclet.java]: Yes, removed.

File: src/main/java/org/broadinstitute/barclay/help/DocWorkUnitHandler.java [hunk 17]
```diff
@@ -0,0 +1,113 @@
+/*
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/barclay/help/DocWorkUnitHandler.java]: Good point. I'm removing from all of the doc files in this PR.
